File: bindings/csharp/AffineUI/Components.cs
// Strongly-typed components — a faithful C# port of include/affineui/components.h
// semantics, implemented over Widget attribute access plus
// affineui_widget_get_kind (typed components deliberately get no C entry
// points; the attribute names are part of the framework contract).
//
// HARD-TO-CRASH CONTRACT (mirrors the C++ header):
//   - reads return a default/empty/typed-zero value,
//   - writes no-op,
//   - IsValid reports whether the component still resolves to a node of the
//     expected kind.
// Three states (Validity): a WrongType component stays attached to whatever
// node it found — generic attribute/text access still works so you can
// inspect it — but all typed accessors are inert. Nothing ever crashes.

using System.Globalization;

namespace AffineUI;

/// <summary>
/// Base for every typed component: holds the underlying <see cref="Widget"/>
/// plus a validity tag and forwards the graceful-degradation surface.
/// </summary>
public abstract class Component
{
    /// <summary>The underlying widget handle (escape hatch for anything not
    /// yet typed).</summary>
    public Widget Widget { get; }

    /// <summary>How this component bound to its node.</summary>
    public Validity Validity { get; }

    private protected Component(Widget widget, Func<WidgetKind, bool> matches)
    {
        Widget = widget;
        WidgetKind kind = widget.Kind;
        Validity = kind == WidgetKind.None ? Validity.NotPresent
                 : matches(kind) ? Validity.Valid
                 : Validity.WrongType;
    }

    /// <summary>True only when bound to a live node of the expected type.</summary>
    public bool IsValid => Validity == AffineUI.Validity.Valid && Widget.IsValid;

    /// <summary>True if some node is attached, even if the wrong type —
    /// useful for diagnosing a <see cref="AffineUI.Validity.WrongType"/>
    /// component.</summary>
    public bool IsAttached => Widget.IsValid;

    /// <summary>The component's stable id (widget-name), or empty.</summary>
    public string Id => Widget.Name;

    /// <summary>The actual widget kind of the attached node (for diagnosing
    /// WrongType), or <see cref="WidgetKind.None"/>.</summary>
    public WidgetKind Kind => Widget.Kind;

    // ── Generic element operations ───────────────────────────────────────
    // NOT type-specific, so they work whenever a node is attached — including
    // in WrongType mode. Type-specific accessors on the derived classes
    // require IsValid instead.

    public string Attr(string name, string fallback = "") => Widget.Attr(name, fallback);

    public void SetAttr(string name, string value) => Widget.SetAttr(name, value);

    public string Text
    {
        get => Widget.Text;
        set => Widget.SetText(value);
    }

    public bool Visible
    {
        get => !Widget.HasAttr("hidden");
        set
        {
            if (value) Widget.RemoveAttr("hidden");
            else Widget.SetAttr("hidden", "");
        }
    }
}

/// <summary>A push button.</summary>
public sealed class Button : Component
{
    internal Button(Widget widget) : base(widget, static k => k == WidgetKind.Button) { }

    public string Label
    {
        get => IsValid ? Widget.Text : "";
        set { if (IsValid) Widget.SetText(value); }
    }

    public bool Enabled
    {
        get => IsValid && !Widget.HasAttr("disabled");
        set
        {
            if (!IsValid) return;
            if (value) Widget.RemoveAttr("disabled");
            else Widget.SetAttr("disabled", "");
        }
    }

    public Button OnClick(Action handler)
    {
        if (IsValid) Widget.OnClick(handler);
        return this;
    }
}

/// <summary>A checkbox / toggle. State is the <c>aria-checked</c> attribute
/// the Decius and Bootstrap themes both read.</summary>
public sealed class Checkbox : Component
{
    internal Checkbox(Widget widget) : base(widget, static k => k == WidgetKind.Checkbox) { }

    public bool Checked
    {
        get => IsValid && Widget.Attr("aria-checked") == "true";
        set { if (IsValid) Widget.SetAttr("aria-checked", value ? "true" : "false"); }
    }

    public Checkbox OnChange(Action<string> handler)
    {
        if (IsValid) Widget.OnChange(handler);
        return this;
    }
}

/// <summary>A single-line text/number/etc. input. <see cref="Value"/> reads
/// the live <c>value</c> attribute.</summary>
public sealed class TextField : Component
{
    internal TextField(Widget widget) : base(widget, static k => k == WidgetKind.TextInput) { }

    public string Value
    {
        get => IsValid ? Widget.Attr("value") : "";
        set { if (IsValid) Widget.SetAttr("value", value); }
    }

    public TextField OnChange(Action<string> handler)
    {
        if (IsValid) Widget.OnChange(handler);
        return this;
    }
}

/// <summary>A dropdown / select. <see cref="Selected"/> reads the chosen
/// option value (<c>data-value</c>).</summary>
public sealed class Dropdown : Component
{
    internal Dropdown(Widget widget) : base(widget, static k => k == WidgetKind.Dropdown) { }

    public string Selected
    {
        get => IsValid ? Widget.Attr("data-value") : "";
        set { if (IsValid) Widget.SetAttr("data-value", value); }
    }

    public Dropdown OnChange(Action<string> handler)
    {
        if (IsValid) Widget.OnChange(handler);
        return this;
    }
}

/// <summary>A slider / range. The group node carries the a11y value
/// (<c>aria-valuenow</c>); the inner control mirrors it in
/// <c>data-value</c>.</summary>
public sealed class Slider : Component
{
    internal Slider(Widget widget) : base(widget, static k => k == WidgetKind.Slider) { }

    /// <summary>Current value, or 0 when unresolved.</summary>
    public double Value => GetValue(0.0);

    /// <summary>Current value, or <paramref name="fallback"/> when the node
    /// is gone or carries no value.</summary>
    public double GetValue(double fallback)
    {
        if (!IsValid) return fallback;
        string text = Widget.Attr("aria-valuenow");
        if (text.Length == 0) text = Widget.Attr("data-value");
        if (text.Length == 0) return fallback;
        return double.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out double v)
            ? v : fallback;
    }

    public Slider OnChange(Action<string> handler)
    {
        if (IsValid) Widget.OnChange(handler);
        return this;
    }
}

/// <summary>A color field/swatch that opens a picker popup.
/// <see cref="Color"/> reads the current CSS color string.</summary>
public sealed class ColorField : Component
{
    // A color field is an input (its builder sets type=color) or a swatch
    // container carrying the color picker toggle.
    internal ColorField(Widget widget)
        : base(widget, static k => k is WidgetKind.TextInput or WidgetKind.Container) { }

    public string Color
    {
        get
        {
            if (!IsValid) return "";
            string v = Widget.Attr("data-value");
            if (v.Length == 0) v = Widget.Attr("value");
            return v;
        }
        set { if (IsValid) Widget.SetAttr("data-value", value); }
    }

    public ColorField OnChange(Action<string> handler)
    {
        if (IsValid) Widget.OnChange(handler);
        return this;
    }
}

/// <summary>A dockable panel. Typed surface for the active tab; visibility
/// uses the generic <see cref="Component.Visible"/>. Tabs/splitters are
/// driven by the interaction layer.</summary>
public sealed class DockPanel : Component
{
    internal DockPanel(Widget widget)
        : base(widget, static k => k is WidgetKind.Container or WidgetKind.Panel or WidgetKind.Card) { }

    public string ActiveTab
    {
        get => IsValid ? Widget.Attr("data-active-tab") : "";
        set { if (IsValid) Widget.SetAttr("data-active-tab", value); }
    }
}

/// <summary>A foldout / collapsible section (<c>aria-expanded</c>).</summary>
public sealed class Foldout : Component
{
    internal Foldout(Widget widget) : base(widget, static k => k == WidgetKind.Container) { }

    public bool Open
    {
        get => IsValid && Widget.Attr("aria-expanded") != "false";
        set { if (IsValid) Widget.SetAttr("aria-expanded", value ? "true" : "false"); }
    }
}

File: bindings/rust/affineui/src/components.rs
//! Strongly-typed components over [`Widget`] — the Rust port of the C++
//! `components.h` semantics (implemented wrapper-side per
//! docs/LANGUAGE_BINDINGS.md §3.7: the attribute names are part of the
//! framework contract, no C entry points needed).
//!
//! Hard-to-crash contract, mirrored exactly:
//! - a query for a missing id yields `Validity::NotPresent`;
//! - a query that finds a node of the WRONG kind stays *attached*
//!   (generic attr/text/visible ops still work) but is not valid, and
//!   all typed accessors are inert;
//! - reads on an invalid component return defaults, writes no-op.

use crate::view::{View, Widget, WidgetKind};

/// Validity of a typed component wrapper (mirrors
/// `affineui::ComponentValidity`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Validity {
    /// Bound to a live node of the expected kind.
    Valid,
    /// Bound to a live node, but NOT this component's kind.
    WrongType,
    /// No node with that id resolves.
    NotPresent,
}

macro_rules! component {
    ($(#[$doc:meta])* $name:ident, matches = $matches:expr) => {
        $(#[$doc])*
        pub struct $name {
            widget: Widget,
            validity: Validity,
        }

        impl $name {
            /// Wrap a queried widget, classifying its validity.
            pub fn from_widget(widget: Widget) -> $name {
                let matches: fn(WidgetKind) -> bool = $matches;
                let validity = if !widget.is_valid() {
                    Validity::NotPresent
                } else if matches(widget.kind()) {
                    Validity::Valid
                } else {
                    Validity::WrongType
                };
                $name { widget, validity }
            }

            pub(crate) fn query(view: &View, name: &str) -> $name {
                Self::from_widget(view.find_widget(name))
            }

            /// True only when bound to a live node of the expected kind.
            pub fn is_valid(&self) -> bool {
                self.validity == Validity::Valid && self.widget.is_valid()
            }

            pub fn validity(&self) -> Validity {
                self.validity
            }

            /// True if some node is attached (even if the wrong kind).
            pub fn attached(&self) -> bool {
                self.widget.is_valid()
            }

            /// The component's stable id (widget key), or empty.
            pub fn id(&self) -> String {
                self.widget.name()
            }

            /// The actual kind of the attached node (for diagnosing
            /// WrongType), or `WidgetKind::None`.
            pub fn kind(&self) -> WidgetKind {
                self.widget.kind()
            }

            // Generic element ops: work whenever a node is attached,
            // including WrongType (components.h contract).

            pub fn attr(&self, name: &str, fallback: &str) -> String {
                self.widget.attr(name, fallback)
            }
            pub fn set_attr(&self, name: &str, value: &str) -> &Self {
                self.widget.set_attr(name, value);
                self
            }
            pub fn text(&self) -> String {
                self.widget.text()
            }
            pub fn set_text(&self, value: &str) -> &Self {
                self.widget.set_text(value);
                self
            }
            pub fn visible(&self) -> bool {
                !self.widget.has_attr("hidden")
            }
            pub fn set_visible(&self, on: bool) -> &Self {
                if on {
                    self.widget.remove_attr("hidden");
                } else {
                    self.widget.set_attr("hidden", "");
                }
                self
            }

            /// Escape hatch to the underlying widget handle.
            pub fn widget(&self) -> &Widget {
                &self.widget
            }
        }
    };
}

component!(
    /// A push button.
    Button,
    matches = |k| k == WidgetKind::Button
);

impl Button {
    pub fn label(&self) -> String {
        if self.is_valid() { self.widget.text() } else { String::new() }
    }
    pub fn set_label(&self, label: &str) -> &Self {
        if self.is_valid() {
            self.widget.set_text(label);
        }
        self
    }
    pub fn enabled(&self) -> bool {
        self.is_valid() && !self.widget.has_attr("disabled")
    }
    pub fn set_enabled(&self, on: bool) -> &Self {
        if self.is_valid() {
            if on {
                self.widget.remove_attr("disabled");
            } else {
                self.widget.set_attr("disabled", "");
            }
        }
        self
    }
    pub fn on_click(&self, f: impl FnMut() + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_click(f);
        }
        self
    }
}

component!(
    /// A checkbox / toggle. State is `aria-checked`.
    Checkbox,
    matches = |k| k == WidgetKind::Checkbox
);

impl Checkbox {
    pub fn checked(&self) -> bool {
        self.is_valid() && self.widget.attr("aria-checked", "") == "true"
    }
    pub fn set_checked(&self, on: bool) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("aria-checked", if on { "true" } else { "false" });
        }
        self
    }
    pub fn on_change(&self, f: impl FnMut(&str) + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_change(f);
        }
        self
    }
}

component!(
    /// A single-line text-like input.
    TextField,
    matches = |k| k == WidgetKind::TextInput
);

impl TextField {
    pub fn value(&self) -> String {
        if self.is_valid() { self.widget.attr("value", "") } else { String::new() }
    }
    pub fn set_value(&self, value: &str) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("value", value);
        }
        self
    }
    pub fn on_change(&self, f: impl FnMut(&str) + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_change(f);
        }
        self
    }
}

component!(
    /// A dropdown / select.
    Dropdown,
    matches = |k| k == WidgetKind::Dropdown
);

impl Dropdown {
    pub fn selected(&self) -> String {
        if self.is_valid() { self.widget.attr("data-value", "") } else { String::new() }
    }
    pub fn set_selected(&self, value: &str) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("data-value", value);
        }
        self
    }
    pub fn on_change(&self, f: impl FnMut(&str) + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_change(f);
        }
        self
    }
}

component!(
    /// A slider / range (value read from `aria-valuenow` / `data-value`).
    Slider,
    matches = |k| k == WidgetKind::Slider
);

impl Slider {
    pub fn value(&self, fallback: f64) -> f64 {
        if !self.is_valid() {
            return fallback;
        }
        let mut text = self.widget.attr("aria-valuenow", "");
        if text.is_empty() {
            text = self.widget.attr("data-value", "");
        }
        text.parse().unwrap_or(fallback)
    }
    pub fn on_change(&self, f: impl FnMut(&str) + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_change(f);
        }
        self
    }
}

component!(
    /// A color field/swatch that opens a picker.
    ColorField,
    matches = |k| k == WidgetKind::TextInput || k == WidgetKind::Container
);

impl ColorField {
    pub fn color(&self) -> String {
        if !self.is_valid() {
            return String::new();
        }
        let v = self.widget.attr("data-value", "");
        if v.is_empty() { self.widget.attr("value", "") } else { v }
    }
    pub fn set_color(&self, css_color: &str) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("data-value", css_color);
        }
        self
    }
    pub fn on_change(&self, f: impl FnMut(&str) + 'static) -> &Self {
        if self.is_valid() {
            self.widget.on_change(f);
        }
        self
    }
}

component!(
    /// A dockable panel (active tab; visibility via the generic ops).
    DockPanel,
    matches = |k| {
        k == WidgetKind::Container || k == WidgetKind::Panel || k == WidgetKind::Card
    }
);

impl DockPanel {
    pub fn active_tab(&self) -> String {
        if self.is_valid() { self.widget.attr("data-active-tab", "") } else { String::new() }
    }
    pub fn set_active_tab(&self, tab_id: &str) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("data-active-tab", tab_id);
        }
        self
    }
}

component!(
    /// A foldout / collapsible section.
    Foldout,
    matches = |k| k == WidgetKind::Container
);

impl Foldout {
    pub fn open(&self) -> bool {
        self.is_valid() && self.widget.attr("aria-expanded", "") != "false"
    }
    pub fn set_open(&self, on: bool) -> &Self {
        if self.is_valid() {
            self.widget.set_attr("aria-expanded", if on { "true" } else { "false" });
        }
        self
    }
}

// ── Typed queries on View ────────────────────────────────────────────

impl View {
    pub fn button_at(&self, name: &str) -> Button {
        Button::query(self, name)
    }
    pub fn checkbox_at(&self, name: &str) -> Checkbox {
        Checkbox::query(self, name)
    }
    pub fn text_field_at(&self, name: &str) -> TextField {
        TextField::query(self, name)
    }
    pub fn dropdown_at(&self, name: &str) -> Dropdown {
        Dropdown::query(self, name)
    }
    pub fn slider_at(&self, name: &str) -> Slider {
        Slider::query(self, name)
    }
    pub fn color_field_at(&self, name: &str) -> ColorField {
        ColorField::query(self, name)
    }
    pub fn dock_panel_at(&self, name: &str) -> DockPanel {
        DockPanel::query(self, name)
    }
    pub fn foldout_at(&self, name: &str) -> Foldout {
        Foldout::query(self, name)
    }
}
